### PR TITLE
feat: Example of event trading framework using client.

### PR DESCRIPTION
**Pull request by adamamsmith**
Demonstrate how to build an event based system around the client in `rubi-py`.

The most important components of this are:

- the BaseEventTradingFramework which is the main loop of and just listens to and directs messages off the queue
- the EventQueue's which ensure concurrent handling of events
- and the ThreadedTransactionManager which allows placement of transactions in a concurrent manner

Still needs a lot of cleanup atm.

### CODE REVIEW

**Comment by adamamsmith** [comment]
Note that this changes a few things in the `rubi` client @denverbaumgartner @ishandhanani :

1. A bug on `place_market_order` with worse_execution_price. The price wasn't multipleid by the order size which means most market orders failed

2. An issue in `_start_default_event_poller` where the node deleted the filter and the poller continually errored but didn't do anything about it or notify the framework. Now the poller will recreate the filter.

3. Some tinkering with the gas values of various methods on `market.py`

4. Finally the client no longer throws and exception if a transaction fails but rather returns the transaction result. This allows better handling of failures where a trading framework can see the failure and then respond instead of crashing (up for more convo on this though)

Otherwise everything else is just additions!

File: bots/example_bots/helpers/grid.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+class Grid:
+
+    def __init__(
+        self,
+        # Inventory
+        starting_base_asset_amount: Decimal,
+        starting_quote_asset_amount: Decimal,
+        starting_base_asset_average_price: Optional[Decimal],
+        # Grid
+        fair_price: Decimal,
```
Comment (denverbaumgartner) [inline bots/example_bots/helpers/grid.py:35]: discussed offline with @adamamsmith - the ability to update `fair_price`, with the necessary actions to happen when updated, will make this a very modular order management system that can be utilized in a variety ways!

File: bots/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+python-dotenv = "1.0.0"
+pyyaml = "6.0.0"
+#rubi = "2.0.6"
+
+[build-system]
+requires = ["poetry-core"]
+build-backend = "poetry.core.masonry.api"
+
+[tool.poetry.group.dev.dependencies]
+rubi = { path = "../rubi", develop = true }
```
Comment (denverbaumgartner) [inline bots/pyproject.toml:19]: new poetry trick for me, this is niccce

File: rubi/network_config/optimism_goerli/network.yaml [hunk 3]
```diff
@@ -15,3 +15,4 @@ token_addresses:
  WETH: "0x4200000000000000000000000000000000000006"
  OP: "0xCeE7148028Ff1B08163343794E85883174a61393"
  USDC: "0xe432f229521eE954f80C83257485405E3d848d17"
+ USDT: "0xd70734ba8101ec28b38ab15e30dc9b60e3c6f433"
```
Comment (denverbaumgartner) [inline rubi/network_config/optimism_goerli/network.yaml:18]: confirmed, `USDT` address supported by the app. can't wait till we have a token list, will be a programmatic way to approach token validation. can even be a part of the CI pipeline

File: rubi/rubi/client.py [hunk 4]
```diff
@@ -190,15 +190,15 @@ def update_pair_allowance(
         """
         pair = self.get_pair(pair_name=pair_name)
 
-        if new_base_asset_allowance and pair.current_base_asset_allowance != new_base_asset_allowance:
+        if new_base_asset_allowance is not None and pair.current_base_asset_allowance != new_base_asset_allowance:
```
Comment (denverbaumgartner) [inline rubi/rubi/client.py:193]: this is just a note for @denverbaumgartner for later, #32 is still outstanding

File: rubi/rubi/contracts/contract_types/transaction_reciept.py [hunk 5]
```diff
@@ -22,7 +23,7 @@ def __init__(
         l1_fee: Optional[int] = None,
         l1_gas_price: Optional[int] = None,
         l1_gas_used: Optional[int] = None,
-        l1_fee_scalar: Optional[float] = None
+        l1_fee_scalar: Optional[Decimal] = None
```
Comment (denverbaumgartner) [inline rubi/rubi/contracts/contract_types/transaction_reciept.py:26]: decimal > float here, agreed 🤝

File: rubi/rubi/contracts/erc20.py [hunk 6]
```diff
@@ -404,3 +404,14 @@ def to_integer(self, number: Decimal) -> int:
...
             return int(number * (10 ** self.decimal))
+
+    def max_approval_amount(self) -> Decimal:
+        """return the max uint256 token approval amount. Note: this is not very secure and if you give this approval you
+        should revoke it when you are done!
+
+        :return: the max approval amount of a uint256 token in Decimal representation
+        :rtype: Decimal
+        """
+        max_uint256 = "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"
```
Comment (denverbaumgartner) [inline rubi/rubi/contracts/erc20.py:415]: this totally works, was wondering if there is a reason we didn't do something like: 
`return self.to_decimal(number=2**256 - 1)`

File: rubi/rubi/contracts/market.py [hunk 7]
```diff
@@ -418,7 +418,7 @@ def batch_requote(
         buy_amts: List[int],
         buy_gems: List[ChecksumAddress],
         nonce: Optional[int] = None,
-        gas: int = 3500000,
+        gas: int = 800000,
```
Comment (denverbaumgartner) [inline rubi/rubi/contracts/market.py:421]: just a cc to: #6

**Review by denverbaumgartner — APPROVED** [review]
this is a really solid framework to build off of, had a good conversation offline with @adamamsmith about this and possible next steps. will turn some of the major points into tickets

**Pull request merged**